xacro_add_files fails if another file in the current building package is included in the file

The cmake function `xacro_add_files` fails when the target `xacro` file has include tags which has the path to files in the current package.
It looks this is because we cannot find the current building package using `ament_index` system.
Is there any fix or workaround?

Here is how to reproduce the issue:

Given a `test_pkg` in a workspace:

- src/test_pkg/CMakeLists.txt

```cmake
cmake_minimum_required(VERSION 3.8)
project(test_pkg)

find_package(ament_cmake REQUIRED)
find_package(xacro REQUIRED)

xacro_add_files(foo.urdf.xacro TARGET foo)

ament_package()
```

- src/test_pkg/foo.urdf.xacro

```xml
<?xml version="1.0" encoding="utf-8"?>
<robot name="robot" xmlns:xacro="http://www.ros.org/wiki/xacro">
  <xacro:include filename="$(find test_pkg)/bar.urdf.xacro"/>
</robot>
```

- src/test_pkg/bar.urdf.xacro

```xml
<?xml version="1.0" encoding="utf-8"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro">
</robot>
```

Run `colcon build` fails on error:

```
$ source /opt/ros/humble/setup.bash
$ colcon build --symlink-install
Starting >>> test_pkg
--- stderr: test_pkg                              
CMake Warning at /opt/ros/humble/share/xacro/cmake/xacro-extras.cmake:74 (message):
  failed to determine deps for: foo.urdf.xacro

  error: package not found: "package 'test_pkg' not found, searching:
  ['/opt/ros/humble']"

  when processing file: foo.urdf.xacro

Call Stack (most recent call first):
  /opt/ros/humble/share/xacro/cmake/xacro-extras.cmake:133 (xacro_add_xacro_file)
  CMakeLists.txt:12 (xacro_add_files)


package not found: "package 'test_pkg' not found, searching: ['/opt/ros/humble']"
when processing file: foo.urdf.xacro
gmake[2]: *** [CMakeFiles/test_pkg_foo.dir/build.make:73: foo.urdf] Error 2
gmake[1]: *** [CMakeFiles/Makefile2:137: CMakeFiles/test_pkg_foo.dir/all] Error 2
gmake: *** [Makefile:146: all] Error 2
---
Failed   <<< test_pkg [0.47s, exited with code 2]

Summary: 0 packages finished [0.57s]
  1 package failed: test_pkg
  1 package had stderr output: test_pkg
```

--- a/cmake/xacro-extras.cmake
+++ b/cmake/xacro-extras.cmake
@@ -46,7 +46,6 @@ function(xacro_add_xacro_file input)
       message(FATAL_ERROR "no <output> specified for: " ${input})
     endif()
   endif()
-  # message(STATUS "output: ${output}")
 
   ## determine absolute output target location
   if(IS_ABSOLUTE ${output})
@@ -54,7 +53,6 @@ function(xacro_add_xacro_file input)
   else()
     set(abs_output ${CMAKE_CURRENT_BINARY_DIR}/${output})
   endif()
-  # message(STATUS "abs_output: ${abs_output}")
 
   ## export abs_output to parent scope in variable ${_XACRO_OUTPUT}
   if(NOT _XACRO_OUTPUT)
@@ -132,7 +130,6 @@ function(xacro_add_files)
 
   # have INSTALL option, but no TARGET: fallback to default target
   if(_XACRO_INSTALL AND NOT _XACRO_TARGET)
-    # message(STATUS "xacro: no TARGET specified, using default")
     set(_XACRO_TARGET _xacro_auto_generate)
   endif()
 

--- a/cmake/xacro-extras.cmake
+++ b/cmake/xacro-extras.cmake
@@ -77,9 +77,16 @@ ${_xacro_err}")
 
   separate_arguments(_xacro_deps_result)
 
+  ## HACK: ament package resolution doesn't work at build time yet
+  # - Augment AMENT_PREFIX_PATH to include ${PROJECT_BINARY_DIR}/ament_cmake_index
+  # - Create a symlink from there to the actual source directory to find source files
+  set(PROJECT_BUILD_INDEX "${PROJECT_BINARY_DIR}/ament_cmake_index")
+  execute_process(COMMAND ${CMAKE_COMMAND} -E make_directory "${PROJECT_BUILD_INDEX}/share")
+  execute_process(COMMAND ${CMAKE_COMMAND} -E create_symlink "${PROJECT_SOURCE_DIR}" "${PROJECT_BUILD_INDEX}/share/${PROJECT_NAME}")
+
   ## command to actually call xacro
   add_custom_command(OUTPUT ${output}
-    COMMAND xacro -o ${abs_output} ${input} ${_XACRO_REMAP}
+    COMMAND ${CMAKE_COMMAND} -E env AMENT_PREFIX_PATH="${PROJECT_BUILD_INDEX}:${AMENT_PREFIX_PATH}" xacro -o ${abs_output} ${input} ${_XACRO_REMAP}
     DEPENDS ${input} ${_xacro_deps_result} ${_XACRO_DEPENDS}
     WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
     COMMENT "xacro: generating ${output} from ${input}"

--- a/cmake/xacro-extras.cmake
+++ b/cmake/xacro-extras.cmake
@@ -85,8 +85,9 @@ ${_xacro_err}")
   execute_process(COMMAND ${CMAKE_COMMAND} -E create_symlink "${PROJECT_SOURCE_DIR}" "${PROJECT_BUILD_INDEX}/share/${PROJECT_NAME}")
 
   ## command to actually call xacro
-  add_custom_command(OUTPUT ${output}
-    COMMAND ${CMAKE_COMMAND} -E env AMENT_PREFIX_PATH="${PROJECT_BUILD_INDEX}:${AMENT_PREFIX_PATH}" xacro -o ${abs_output} ${input} ${_XACRO_REMAP}
+  list(JOIN AMENT_PREFIX_PATH ":" AMENT_PREFIX_PATH_ENV) # format as colon-separated list
+  add_custom_command(OUTPUT ${abs_output}
+    COMMAND ${CMAKE_COMMAND} -E env AMENT_PREFIX_PATH="${PROJECT_BUILD_INDEX}:${AMENT_PREFIX_PATH_ENV}" xacro -o ${abs_output} ${input} ${_XACRO_REMAP}
     DEPENDS ${input} ${_xacro_deps_result} ${_XACRO_DEPENDS}
     WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
     COMMENT "xacro: generating ${output} from ${input}"

